File: logic/socialize.py
import re

from logic import content_remote
from logic import pubsubhubbub_publish
from logic import users as user_logic

def socialize(handler, content):
  if content.hidden:
    return

  # TODO support @blah@blah.com
  mentions  = re.findall(r'@(\w+)', content.view, re.M | re.U)
  mentions += re.findall(r'\+(\w+)', content.view, re.M | re.U)

  reply(handler, content, mentions)

  publish(handler, content)

def publish(handler, content):
  try:
    pubsubhubbub_publish.publish(handler.constants['push_hub'],
        handler.nav_url(host=True, username=content.username, section='feed'))
  except pubsubhubbub_publish.PublishError, e:
    # do nothing for now
    pass

def reply(handler, content, mentions=None):
  profile = content.username

  thread_user_remote = None
  users = []
  users_profile = []
  mentioned_users = []
  if content.thread:
    thread_content = handler.models.content_remote.get(to_username=profile,
        post_id=content.thread)[0]
    if not thread_content:
      return
    thread_user_remote = handler.models.users_remote.get(
        local_username=profile, profile_url=thread_content.from_user)[0]

    if thread_user_remote and thread_user_remote.webmention_url:
      users.append(thread_user_remote)
      users_profile.append(thread_user_remote.profile_url)
      mentioned_users.append(thread_user_remote)

  if mentions:
    for user in mentions:
      user_remote = handler.models.users_remote.get(local_username=profile,
          username=user)[0]

      if not user_remote or (thread_user_remote and
          thread_user_remote.profile_url == user_remote.profile_url):
        continue

      if user_remote.profile_url in users_profile:
        continue

      if user_remote.webmention_url:
        users.append(user_remote)
        users_profile.append(user_remote.profile_url)
        mentioned_users.append(user_remote)

  if content.comments_count:
    comments = content_remote.get_comments(handler, content)

    for comment in comments:
      user_remote = handler.models.users_remote.get(local_username=profile,
          username=comment.from_user)[0]

      if not user_remote:
        continue

      if user_remote.profile_url in users_profile:
        continue

      # TODO(mime): Necessary to check webmention_url?
      if user_remote and user_remote.webmention_url:
        users.append(user_remote)
        users_profile.append(user_remote.profile_url)

  for user in users:
    user_logic.webmention_reply(handler, user, content, thread=content.thread,
        mentioned_users=mentioned_users)

File: server/api/social_butterfly/socialize.js
import constants from '../../../shared/constants';
import { contentUrl } from '../../../shared/util/url_factory';
import { convertFromRaw } from 'draft-js';
import { comment as emailComment } from './email';
import fetch from 'node-fetch';
import models from '../../data/models';
import { reply as salmonReply } from './salmon';
import { webmentionReply } from './webmention';

export default async function socialize(req, contentOwner, localContent, opt_remoteContent, opt_isComment) {
  if (localContent.hidden) {
    return;
  }

  await pubsubhubbubPush(req, localContent);

  await parseMentions(req, contentOwner, localContent, opt_remoteContent);

  if (opt_isComment) {
    const localContentUser = await models.User.findOne({ where: { username: localContent.username } });
    emailComment(
      req,
      opt_remoteContent.username,
      opt_remoteContent.comment_user,
      localContentUser.email,
      contentUrl(localContent, req),
      opt_remoteContent
    );
  }
}

const MENTION_REGEX = /[@+](\w+)/g;
async function parseMentions(req, contentOwner, content, opt_remoteContent) {
  const remoteUsers = [];
  const mentionedRemoteUsers = [];

  function addToUsersList(userRemote, shouldAddToMentions) {
    if (!userRemote || remoteUsers.find(el => el.profile_url === userRemote.profile_url)) {
      return;
    }

    if (userRemote.webmention_url) {
      remoteUsers.push(userRemote);
      shouldAddToMentions && mentionedRemoteUsers.push(userRemote);
    }
  }

  // Add user from original thread as mentioned user.
  if (content.thread) {
    const threadContent = await models.Content_Remote.findOne({
      where: { to_username: content.username, post_id: content.thread },
    });
    if (!threadContent) {
      return;
    }
    const threadUserRemote = await models.User_Remote.findOne({
      where: { local_username: content.username, profile_url: threadContent.from_user },
    });
    addToUsersList(threadUserRemote, true);
  }

  // Find mentions in the text.
  // TODO(mime): support @blah@blah.com
  const plaintext = content.content ? convertFromRaw(JSON.parse(content.content)).getPlainText() : content.view;
  const mentions = plaintext.match(MENTION_REGEX) || [];
  for (const mention of mentions) {
    const userRemote = await models.User_Remote.findOne({
      where: { local_username: content.username, username: mention.slice(1) },
    });
    addToUsersList(userRemote, true);
  }

  // Find all users from the comments.
  if (content.comments_count) {
    const comments = await models.Content_Remote.findAll({
      where: {
        to_username: content.username,
        local_content_name: content.name,
        type: 'comment',
      },
    });
    for (const comment of comments) {
      const userRemote = await models.User_Remote.findOne({
        where: { local_username: content.username, username: comment.from_user },
      });
      addToUsersList(userRemote, false);
    }
  }

  if (opt_remoteContent) {
    // Find mentions in the comment text.
    // TODO(mime): support @blah@blah.com
    const plaintext = convertFromRaw(JSON.parse(opt_remoteContent.content)).getPlainText();
    const mentions = plaintext.match(MENTION_REGEX) || [];
    for (const mention of mentions) {
      const userRemote = await models.User_Remote.findOne({
        where: { local_username: content.username, username: mention.slice(1) },
      });
      addToUsersList(userRemote, true);
    }
  }

  for (const userRemote of remoteUsers) {
    webmentionReply(req, userRemote, content, content.thread, mentionedRemoteUsers);
    salmonReply(req, contentOwner, content, userRemote.salmon_url, mentionedRemoteUsers);
  }
}

async function pubsubhubbubPush(req, content) {
  try {
    const contentFeedUrl = contentUrl(content, req);
    await fetch(constants.pushHub, {
      method: 'POST',
      body: new URLSearchParams({ 'hub.url': contentFeedUrl, 'hub.mode': 'publish' }),
    });
  } catch (ex) {
    // Not a big deal if this fails.
  }
}
